modules/criteoBidAdapter.js

```diff
@@ -0,0 +1,279 @@
+        } else {
+          bid.ad = slot.creative;
+        }
+        bids.push(bid);
+      });
+    }
+
+    return bids;
+  },
+};
+
+/**
+ * @return {boolean}
+ */
+function publisherTagAvailable() {
+  return typeof Criteo !== 'undefined' && Criteo.PubTag && Criteo.PubTag.Adapters && Criteo.PubTag.Adapters.Prebid;
+}
+
+/**
+ * @param {BidRequest[]} bidRequests
+ * @return {CriteoContext}
+ */
+function buildContext(bidRequests) {
+  const url = utils.getTopWindowUrl();
+  const queryString = parse(url).search;
+
+  const context = {
+    url: url,
+    debug: queryString['pbt_debug'] === '1',
+    noLog: queryString['pbt_nolog'] === '1',
+    integrationMode: undefined,
+  };
+
+  bidRequests.forEach(bidRequest => {
+    if (bidRequest.params.integrationMode) {
+      context.integrationMode = bidRequest.params.integrationMode;
+    }
+  })
+
+  return context;
+}
+
+/**
+ * @param {CriteoContext} context
+ * @return {string}
+ */
+function buildCdbUrl(context) {
+  let url = CDB_ENDPOINT;
+  url += '?profileId=' + PROFILE_ID;
+  url += '&av=' + String(ADAPTER_VERSION);
+  url += '&cb=' + String(Math.floor(Math.random() * 99999999999));
+
+  if (context.integrationMode in INTEGRATION_MODES) {
+    url += '&im=' + INTEGRATION_MODES[context.integrationMode];
+  }
+  if (context.debug) {
+    url += '&debug=1';
+  }
+  if (context.noLog) {
+    url += '&nolog=1';
+  }
+
+  return url;
+}
+
+/**
+ * @param {CriteoContext} context
+ * @param {BidRequest[]} bidRequests
+ * @return {*}
+ */
+function buildCdbRequest(context, bidRequests) {
+  let networkId;
+  const request = {
+    publisher: {
+      url: context.url,
+    },
+    slots: bidRequests.map(bidRequest => {
+      networkId = bidRequest.params.networkId || networkId;
+      const slot = {
+        impid: bidRequest.adUnitCode,
+        transactionid: bidRequest.transactionId,
+        auctionId: bidRequest.auctionId,
+        sizes: bidRequest.sizes.map(size => size[0] + 'x' + size[1]),
+      };
+      if (bidRequest.params.zoneId) {
+        slot.zoneid = bidRequest.params.zoneId;
+      }
+      if (bidRequest.params.publisherSubId) {
+        slot.publishersubid = bidRequest.params.publisherSubId;
+      }
+      if (bidRequest.params.nativeCallback) {
+        slot.native = true;
+      }
+      return slot;
+    }),
+  };
+  if (networkId) {
+    request.publisher.networkid = networkId;
+  }
+  return request;
+}
+
+/**
+ * @param {string} id
+ * @param {*} payload
+ * @param {*} callback
+ * @return {string}
+ */
+function createNativeAd(id, payload, callback) {
+  // Store the callback and payload in a global object to be later accessed from the creative
+  window.criteo_prebid_native_slots = window.criteo_prebid_native_slots || {};
+  window.criteo_prebid_native_slots[id] = { callback, payload };
+
+  // The creative is in an iframe so we have to get the callback and payload
+  // from the parent window (doesn't work with safeframes)
```

Comment:
do you have a postMessage solution that would work in safeFrames?

Reply:
Maybe, we will study this possibility and push it in a second iteration if feasable.

Reply:
that's fine